### PR TITLE
feat: Get trainee TIS ID from the token

### PR DESCRIPTION
**Pull request by Judge40**
Currently the trainee's TIS ID is provided as part of the request URL,
as this needs to be locked down to only allow the logged in user to see
their own details the ID should be taken from the authorization token
instead.

TISNEW-3833

### CODE REVIEW

**Comment by Judge40** [comment]
Created as a draft because there are API Gateway changes that are required when this is merged.

File: src/main/java/uk/nhs/hee/trainee/details/api/TraineeProfileResource.java [hunk 1]
```diff
@@ -48,37 +55,40 @@ public TraineeProfileResource(TraineeProfileService traineeProfileService,
...
+      @RequestHeader(HttpHeaders.AUTHORIZATION) String token) {
+    log.trace("Trainee Profile of authenticated user.");
+
+    String[] tokenSections = token.split("\\.");
+    byte[] payloadBytes = Base64.getDecoder()
+        .decode(tokenSections[1].getBytes(StandardCharsets.UTF_8));
+    String tisId;
+
+    try {
+      ObjectMapper mapper = new ObjectMapper();
```
Comment (JosephKelly) [inline src/main/java/uk/nhs/hee/trainee/details/api/TraineeProfileResource.java:75]: It's looks safe to declare this as a field and avoid creating it per-request.

**Review by jayanta2018 — DISMISSED** [review]

File: src/main/java/uk/nhs/hee/trainee/details/api/TraineeProfileResource.java (hunk 1 above)
Comment (JosephKelly) [reply src/main/java/uk/nhs/hee/trainee/details/api/TraineeProfileResource.java:75]: ...or allowing Dependency Injection for testing?

File: src/main/java/uk/nhs/hee/trainee/details/api/TraineeProfileResource.java [hunk 2]
```diff
@@ -48,37 +55,40 @@ public TraineeProfileResource(TraineeProfileService traineeProfileService,
...
+      tisId = (String) payload.get(TIS_ID_ATTRIBUTE);
+    } catch (IOException e) {
+      log.warn("Unable to read tisId from token.", e);
+      return ResponseEntity.badRequest().build();
+    }
+
+    TraineeProfile traineeProfile = traineeProfileService.getTraineeProfileByTraineeTisId(tisId);
+
+    if (traineeProfile == null) {
+      log.warn("Trainee profile not found for id {}.", tisId);
```
Comment (JosephKelly) [inline src/main/java/uk/nhs/hee/trainee/details/api/TraineeProfileResource.java:86]: Think this is a great example of what we should look out for.

**Review by JosephKelly — COMMENTED** [review]
Looks really good.  1 note to consider whether to make Jackson `ObjectMapper` longer lived.  A quick search suggests it being configured elsewhere & shared is fine too.

File: src/main/java/uk/nhs/hee/trainee/details/api/TraineeProfileResource.java (hunk 1 above)
Comment (Judge40) [reply src/main/java/uk/nhs/hee/trainee/details/api/TraineeProfileResource.java:75]: Good point, I'll inject it 👍

**Review by JosephKelly — APPROVED** [review]

**Pull request merged**